Add EDE and dependencies

kraileth opened the pull request:
Add the following ports:

- Gamin
- FTjam
- Edelib
- EDE

I do have a few questions/comments regarding each port.

Gamin: This required fbsd specific patches. I don't currently have a Linux machine to test it there, but my guess is that it's geared towards Linux and might just work. For dfly I'm not so confident. What is the right approach in this case - mark as "FreeBSD-only" because only that was tested? Or add breakage only information when it's confirmed that it doesn't work on some platform?

FTjam: Might sound redundant since boost-jam is already available. However despite boost-jam's claims to stay compatible with mainline Jam it doesn't live up to that which is why either Jam or the extended FTjam variant is needed to build EDE (and some other applications).

Edelib: Whoever created the FreeBSD port for EDE initially chose to put Edelib and EDE together in just one port. No other project does this, edelib is always packaged separately. I've still set the FPC_EQUIVALENT in the specification since it's part of that referenced port. Is this correct or should I leave it out since there's no fbsd port matching it **exactly**?

EDE: The port is not perfect, yet. Visual glitches occur in the window decoration (I've had this years ago on Arch Linux, too, needs further investigation) and the keyboard tool seems to be unable to detect available keymaps (I assume hard-coded paths to be the problem here). Other than that it seems to work pretty well, providing a simple desktop environment with a panel and a menu, tools like a calendar, window switcher, etc. However I chose to simply deactivate documentation - I didn't feel like porting doxygen and all of its friends right now (clearly need some more experience before even attempting that).

The current status is that the EDE port is definitely usable and provides some value to the user despite the fairly minor glitches. I would like to get this committed now and unhurriedly work out the remaining problems together with the EDE author (will probably take a some weeks of mailing back and forth). Thus it wouldn't be blocking my repo and I can work on other ports in the meantime. Since it generally does work, would you accept the port for now? Chapter 14 on the wiki says "Once you are confident in the quality of the port" - is this satisfied by "it didn't explode on startup, I checked functionality for a couple of minutes and found it to be usable" plus "I personally think it's preferable over Openbox and intend to use it on my Ravenports test machine from now on, trying to improve it over time"?

jrmarino commented:
gamin: rename the patches directory to "freebsd".
"patches" are for patches common to all operating systems.  If a patch is limited to a specific opsys, it goes in freebsd, dragonfly, linux, etc.

jrmarino commented:
Edelib: use FPC_EQUIVALENT on one port only.  On the other port, just don't define it.  It has to be 1:1.

jrmarino commented:
Since you're currently the only user of EDE, a "usable but still work in progress" state of the port is fine.

jrmarino commented:
on ede post-install, having one line per strip command seems excessive.
You can remove many lines with a glob (e..g ede-*)
or you can use a .for/.endfor loop and just iterate (so it's just one command inside a loop).
The result is the same, but the specification is much shorter

jrmarino commented:
if you're reasonably confident that it won't work on dragonfly without specific patches, you can mark it as NOT_FOR_OPSYS=dragonfly.  somebody else can add patches and test if they want it for dragonfly later.

jrmarino commented:
I guess you could also set "BROKEN[dragonfly]".
BROKEN implies temporarily broken (there's an intention to support/fix) whereas NOT_FOR_OPSYS implies a permanent non-support.  Given that, EDE is better marked as BROKEN.

kraileth commented:
I have some questions:

Gamin: The patches directory is already named "freebsd". Or did you mean that the patches in it should be named freebsd-*, too? Probably not.

Edelib: Ok, removed the FPC_EQUIVALENT in my tree.

Ede: Great that you're following a pragmatic approach here! I'm just having trouble with the strip commands. I can easily remove 9 lines by using a glob for the shared objects. The executable files an the bin directory are a problem however. It's mostly ELF binaries, but there are three shell scripts among them which makes the stripping fail for those.

I've tried something like this **(I replaced the backticks with carets since GH uses them for formating)**:

`for f in ^find${STAGEDIR}${PREFIX}/bin/  -type f -iname ede-*^ ; do [ ^file ${f} | grep ELF -c^ -gt 0 ] && echo ${f}`

The problem here is that I cannot get ${f} passed on to the shell that executes the command. I've tried escaping the dollar sign, but to no avail. Is there an easy solution to do this? If not I will work around the problem in a less elegant but also shorter way.

I'll also set gamin as broken on dfly. That sounds like the way to go.

jrmarino commented:
gamin is correct now.  I assumed you had the patches in a common patches directory but didn't check.

the ede post-install looks overly complex.  A hardcoded loop would be clearer.
If the glob catches the wrong files, don't use the glob, just have a very long "for/endfor" loop.

As I said, in practical terms it's no different that what you had, but somehow seems better (which is subjective of course).

kraileth commented:
Ok, everything should now be in better shape.

kraileth commented:
BTW: If I find broken ports that I fail to fix, do you want me to create a pull request marking them BROKEN[someOS] or should I create an issue for it here on GH?

jrmarino commented:
so ede on linux and solaris (which I'm working on bootstrapping) will not depend on gamin?
Or asking another way, why does ede have a dependency on gamin only when run on FreeBSD?
I just don't the background of the conditional dependency here.

kraileth commented:
On Linux EDE does not need gamin, that's correct. It uses inotify directly in that case. About Solaris: Huh, no idea... In general EDE will work without gamin on any platform, but some features won't be available (like updating the menu if software was installed or deinstalled). On FreeBSD gamin makes that feature work - and the same should be true for Linux systems where the kernel was compiled without inotify support. Since it shouldn't do any harm, I'll change it to be a general dependency. That should be the easiest thing.

jrmarino commented:
actually, I was satisfied with your answer.
How about this?
Revert the last change, but add conditional dependencies on "dragonfly" and "sunos".
If we find sunos doesn't use it, we can remove it later.  But let's not add an unnecessary dep for linux users.

kraileth commented:
Ok, sure thing. Committed the proposed changes (looking forward to the new platform BTW. Might be a good reason to finally install OpenIndiana Hipster on true hardware!)..

jrmarino commented:
i just noticed ftjam is a release candidate.
Why was a release candidate taken over the last stable release?
(Yes I know both are from 2006, but rc's need solid justification to be used over a true release)

jrmarino commented:
if there's not a good reason (e.g. "I just used the version FreeBSD had") then we should downgrade to version 2.5.2 and set the EPOCH to 1.

kraileth commented:
FreeBSD actually has the stable version in ports. Arch uses the newer RC - and has always used that for years. I think that it fixed something (I remember problems building EDE on Ubuntu, but since that was nearly five years ago I'm not sure that it was due to the build tool). If I remember the story correctly, it was meant to become the next release anyway and since it implemented something desirable people started using it. But then the developer unexpectedly never found time to do the proper release and so they just stuck with what they had. According to Repology, it's mainly Debian and Ubuntu carrying the old release and a lot of the other repos hold the RC version as well.

There's only a couple of programs that have ftjam as a build dependency in the Arch repo today. If you really think downgrading is necessary I can try to build those with the old version and see if any important feature is missing. If not, my take at a reason would be "that version has been in very wide-spread use for over a decade" in lots of Linux distros, aiming at the popularity of it.

jrmarino commented:
The thing with Arch (and FreeBSD ports for that matter) is that there's no consistent policy.  If a particular package maintainer firmly believes release candidates are better than releases in every case, he/she might immediately change the version and all the users just have to live with it.  So just because it's been there for years doesn't really prove anything to me.

I understand the ftjam situation, and it's what I expected it would be.  Never releasing an RC is a bad sign (theoretically an RC is thought to be release quality and it's just out there for final inspection, so "releasing" would be a matter of changing the file names, so how is it an author can't find time for something that simple?)

It doesn't help that we don't know what this RC fixes and if it's critical.  Maybe what it fixes isn't needed by ede.   By the way, this isn't the first instance of something like this.  I had to downgrade openjade when I realized that FreeBSD was using a pre-release of it.

so my preference is: if ede builds and works fine with the last full release of ftjam (which I assume is just a build dependency) then the best thing to do is "downgrade" ftjam to the last release.  That helps ravenports have a reputation of a consistent quality with regards to software releases.

Conversely, if ede won't build with the last release, now we have a justifiable reason to keep the RC (which should be documented in the description file).  That's my opinion.

jrmarino commented:
by the way, given that ftjam is now 11 years old, how come ede uses it?  Is ede also 11 years old?  or older?
You'd think they would have migrated to something maintained ...

jrmarino commented:
instead of 11 years old, I mean 11 years abandoned

jrmarino commented:
I'm worried about something else now.
ftjam didn't build on DragonFly.  I checked and found there are 19 patches in FreeBSD for release 2.5.2, but this version has no patches.  Was every issue fixed by the 2.5.2 patches resolved in 2.5.3 (other than jam.h which probably explains breakage on DragonFly)?  Or in other words: did you go through any process that ended up intentionally eliminating these patches?

jrmarino commented:
a quick glance seems to indicate most of these patches are fixing ambiguitities and missing headers that older gccs accepted.  Not uncommon on ancient c++ programs.

jrmarino commented:
obviously the quickest solution is drop back down to 2.5.2 and bring in all 19 freebsd patches, but I'm interested in any information you have to share first.

kraileth commented:
No, EDE is not that old - at least not the latest release. The whole project is 20 years old and in fact a few month older than even E17. Version 2.1 was released in mid 2014, development continued (slowly) since then, but another release wasn't cut so far.

FTjam on the other hand really looks like a dead end. I started working with EDE on Arch Linux years around 2012. FTjam is the jam variant that they package, which is why I've used it on other platforms, too. EDE itself needs "any Jam that's not boost jam". Both FreeBSD and OpenBSD use mainline Jam to build their particular ports and in fact EDE offered the Jam source, too, so people could easily build it if not available. I was under the impression that FTjam was superior as it added more features. Now taking a closer look, today that more or less sounds like a joke: Except for a few useful things like regex replacement it mainly adds compatibility to run on Win 9.x (!) and OS/2.

Ironically, after more than 10 years the original Jam saw a new release 2.6 in late 2014. I totally managed to miss that one. Hm. Taking all of that into consideration I would even propose dropping FTjam completely and go for mainline Jam. If you agree that this is the way to go I'll see if I can get a working Jam port, make sure EDE builds with it (which I'm pretty confident about) and then FTjam can go away.

jrmarino commented:
yes, I agree with your proposal.  Let's replace ftjam completely and drop it.

jrmarino commented:
There's something fishy with gamin.
I have it building on DragonFly (with one patch; the same patch could replace the 4 patches on FreeBSD).
However, the description says "It uses kqueue/kevent on FreeBSD and inotify on Linux." but it seems all the patches that make that happen were left out.  On devel/gamin, there are many more patches that did not make it over.
Mistake or intentional?

jrmarino commented:
also, on DragonFly, shared libraries are building but apparently not on FreeBSD??
```
====> Checking for package manifest issues
Orphaned file detected: lib/libfam.so.0.0.0
Orphaned file detected: lib/libgamin-1.so.0
Orphaned file detected: lib/libfam.so
Orphaned file detected: lib/libfam.so.0
Orphaned file detected: lib/libgamin-1.so.0.1.10
Orphaned file detected: lib/libgamin-1.so
```

jrmarino commented:
i see why.  This port needs the freebsd10 fix module.  Means this tarball is really old.

jrmarino commented:
Okay, I fixed some of this (untested on FreeBSD) here: bc2f52ca798243a28545512df28366bf3283b4f8
More work is needed I think (described in commit message)

jrmarino commented:
ah, by the way, try to keep the manifests in alphabetical order.  It helps the "compression" seen on the compiled buildsheet.

jrmarino commented:
another by the way, I realligned ede a few minutes ago: ebd974aa0514e4c9c1e5472ee1394cb9177105dc

kraileth commented:
Thanks for your efforts to support as many platforms as good as possible! Gamin still builds fine after your update. About the patches: I added what was needed to make it compile. I guess I forgot the others then... To be honest, though, I'm not really sure which ones would be needed. "patch-server_gam_kqueue.c" sounds like an excellent candidate here. Should the others be applied, too? If yes I'll add them.

jrmarino commented:
I just realized that the 3 REINPLACE_CMD lines in ede could be replaced by one line.
e.g. instead of
```
	${REINPLACE_CMD} -e "s|__LDFLAGS__|${LDFLAGS}|" \
		${WRKSRC}/build/Program.jam
	${REINPLACE_CMD} -e "s|__LDFLAGS__|${LDFLAGS}|" ${WRKSRC}/evoke/Jamfile
	${REINPLACE_CMD} -e "s|__LDFLAGS__|${LDFLAGS}|" ${WRKSRC}/pekwm/Jamfile
```
it could be:
```
	${REINPLACE_CMD} -e "s|__LDFLAGS__|${LDFLAGS}|" \
		${WRKSRC}/build/Program.jam ${WRKSRC}/evoke/Jamfile ${WRKSRC}/pekwm/Jamfile
```
or
```
	${REINPLACE_CMD} -e "s|__LDFLAGS__|${LDFLAGS}|" \
		${WRKSRC}/build/Program.jam \
		${WRKSRC}/evoke/Jamfile \
		${WRKSRC}/pekwm/Jamfile
```

jrmarino commented:
> To be honest, though, I'm not really sure which ones would be needed. 

Well, it's basically on you to understand *why* each patch is there is determine if it should be in ravenports.  Any patch that is place to support clang can be dropped.  Often patches are obsolete and the original maintainer forgot to remove it.

But most importantly, they are there to fix functionality.  It's more than "it compiles", it is there to make it work.  So you can't forget about freebsd patches at all. 

I'd take another look.  Or at least decide you don't want the modification as an informed decision.

kraileth commented:
Yes, I'll take a closer look at them and try to figure out which ones should be needed.

kraileth commented:
My goodness, that EDE REINPLACE block is terrible indeed. Should really have read over it again (or did I overlook it?). Corrected.

jrmarino commented:
i assumed it was a straight "translation" from freebsd ports.
There are tons of bad styles and straight up mistakes in ports, so really every line should be inspected during a port translation with a critical eye.  Obviously it worked but it nearly always can be done better.
